### PR TITLE
Deserializer: Always validate strings, introduce fromBinary, handle pointers

### PR DESCRIPTION
**Pull request by Geod24**
Yet another deserialization PR. Better reviewed commit-by-commit.

### CODE REVIEW

File: source/agora/common/Serializer.d [hunk 1]
```diff
@@ -380,6 +380,18 @@ public void serializePart (T) (scope const auto ref T record, scope SerializeDg
             entry.serializePart(dg);
     }
 
+    // Pointers are handled as arrays, only their size must be 0 or 1
```
Comment (AndrejMitrovic) [inline source/agora/common/Serializer.d:383]: Recursive / Cyclic pointers are not supported, right? Hopefully we'll never use them though.

Comment (Geod24) [reply source/agora/common/Serializer.d:383]: Good point, I guess that's why they aren't usually supported... And I suppose inner pointers will mess things too :/

Comment (MichaelKim20) [reply source/agora/common/Serializer.d:383]: I understand what situations are used through unittest.

File: source/agora/common/Serializer.d [hunk 2]
```diff
@@ -152,11 +152,17 @@ unittest
             serializePart(this.s, dg);
         }
 
-        void deserialize (scope DeserializeDg dg) @safe
+        static QT fromBinary (QT) (
```
Comment (MichaelKim20) [inline source/agora/common/Serializer.d:155]: We must implement this as static.
And we need to use 'deserializeFull' inside.

Comment (Geod24) [reply source/agora/common/Serializer.d:155]: https://bpfkorea.github.io/agora/agora/common/Serializer.html

**Review by MichaelKim20 — APPROVED** [review]
LGTM

**Review by AndrejMitrovic — APPROVED** [review]

**Pull request merged**